pool.go

```diff
@@ -2,47 +2,111 @@ package valkey
 
 import (
 	"context"
+	"errors"
 	"sync"
 	"time"
 )
 
-func newPool(cap int, dead wire, cleanup time.Duration, minSize int, makeFn func(context.Context) wire) *pool {
+const (
+	PoolTimeoutExceeded = "pool timeout exceeded"
+)
+
+var poolTimeoutError = errors.New(PoolTimeoutExceeded)
+
+func newPool(cap int, dead wire, cleanup time.Duration, minSize int, poolTimeout time.Duration, makeFn func(context.Context) wire) *pool {
 	if cap <= 0 {
 		cap = DefaultPoolSize
 	}
 
 	return &pool{
-		size:    0,
-		minSize: minSize,
-		cap:     cap,
-		dead:    dead,
-		make:    makeFn,
-		list:    make([]wire, 0, 4),
-		cond:    sync.NewCond(&sync.Mutex{}),
-		cleanup: cleanup,
+		size:        0,
+		minSize:     minSize,
+		cap:         cap,
+		dead:        dead,
+		make:        makeFn,
+		list:        make([]wire, 0, 4),
+		cond:        sync.NewCond(&sync.Mutex{}),
+		cleanup:     cleanup,
+		poolTimeout: poolTimeout,
 	}
 }
 
 type pool struct {
-	dead    wire
-	cond    *sync.Cond
-	timer   *time.Timer
-	make    func(ctx context.Context) wire
-	list    []wire
-	cleanup time.Duration
-	size    int
-	minSize int
-	cap     int
-	down    bool
-	timerOn bool
+	dead        wire
+	cond        *sync.Cond
+	timer       *time.Timer
+	make        func(ctx context.Context) wire
+	list        []wire
+	cleanup     time.Duration
+	size        int
+	minSize     int
+	cap         int
+	down        bool
+	timerOn     bool
+	poolTimeout time.Duration
 }
 
 func (p *pool) Acquire(ctx context.Context) (v wire) {
 	p.cond.L.Lock()
+
+	poolDeadline := time.Time{}
+	if p.poolTimeout > 0 {
+		poolDeadline = time.Now().Add(p.poolTimeout)
+	}
+
+	if ctxDeadline, ok := ctx.Deadline(); ok {
+		if poolDeadline.IsZero() {
+			poolDeadline = ctxDeadline
+		} else if ctxDeadline.Before(poolDeadline) {
+			poolDeadline = ctxDeadline
+		}
+	}
+
+	var (
+		poolCtx context.Context
+		cancel  context.CancelFunc
+	)
+	if !poolDeadline.IsZero() {
+		poolCtx, cancel = context.WithDeadline(context.Background(), poolDeadline)
+		defer cancel()
+
+		go func() {
+			<-poolCtx.Done()
+			if poolCtx.Err() == context.DeadlineExceeded { // signal the pool to stop waiting, only if the poolctx is deadline exceeded
+				p.cond.Signal()
```

Comment:
```suggestion
				p.cond.Broadcast()
```
Using Signal seems incorrect. There is no guarantee that Signal will wake the correct goroutine up.

Reply:
@rueian I agree that it doesn’t guarantee waking up the right goroutine, but I suspect that calling `Broadcast()` every time the ctx deadline exceeds—especially if it happens frequently—could increase mutex contention, as all waiting goroutines wake up unnecessarily and compete for the lock. wdyt?

Reply:
Yes, but I am afraid that we have no choice as long as we still use the sync.Cond. If the standard .Signal() wakes a wrong goroutine, which is very likely, then the timeout just doesn’t work.

Reply:
In production throughput, Is it reasonable to assume that if a pool context deadline exceeds, it likely affects multiple simultaneous acquire calls, meaning this particular Wait() will be awakened by one signal or another?